controller/controller_server/csi_controller_server.py

```diff
@@ -318,6 +303,71 @@ def ListVolumes(self, request, context):
         logger.info("finished ListVolumes")
         return csi_pb2.ListVolumesResponse()
 
+    def CreateSnapshot(self, request, context):
+        set_current_thread_name(request.name)
+        try:
+            utils.validate_create_snapshot_request(request)
+        except ValidationException as ex:
+            logger.error("failed request validation")
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            return csi_pb2.CreateSnapshotResponse()
+
+        source_volume_id = request.source_volume_id
+        logger.info("Snapshot base name : {}. Source volume id : {}".format(request.name, source_volume_id))
+        secrets = request.secrets
+        user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+        try:
+            _, vol_id = utils.get_volume_id_info(source_volume_id)
+            with ArrayConnectionManager(user, password, array_addresses) as array_mediator:
+                logger.debug(array_mediator)
+                # TODO: CSI-1358 - remove try/except
+                try:
+                    snapshot_name = self._get_snapshot_name(request, array_mediator)
+                except controller_errors.IllegalObjectName as ex:
+                    context.set_details(ex.message)
+                    context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                    return csi_pb2.CreateSnapshotResponse()
+
+                volume_name = array_mediator.get_volume_name(vol_id)
+                logger.info("Snapshot name : {}. Volume name : {}".format(snapshot_name, volume_name))
+                snapshot = array_mediator.get_snapshot(snapshot_name)
+                if not snapshot:
+                    logger.debug(
+                        "Snapshot doesn't exist. Creating a new snapshot {0} from volume {1}".format(snapshot_name,
+                                                                                                     volume_name))
+                    snapshot = array_mediator.create_snapshot(snapshot_name, volume_name)
```

Comment:
what if there are two create snapshot call? the first is faster so it creates the snapshot, should the second return error?

Reply:
two calls for the same snapshot?
We have the same issue for create volume. In  such case slower create will fall,the retry will occur an d succeed.

Reply:
Ori created ticket to handle this it it will be decided to fix.  CSI-1380.
I don't add TODO unless we'll decide to handle it this way